[Doc] [Jobs] Define driver and unify job definitions (#29608)
ray-project/ray: doc/source/cluster/key-concepts.rst

It is important to understand that the autoscaler only reacts to task and actor resource requests, and not application metrics or physical resource utilization.
To learn more about autoscaling, refer to the user guides for Ray clusters on :ref:`VMs <cloud-vm-index>` and :ref:`Kubernetes <kuberay-index>`.

.. _cluster-clients-and-jobs:

> Ray Job

Reviewer: Should it be `Ray Jobs` as the title?

-------

A Ray job is a single application: it is the collection of Ray tasks, objects, and actors that originate from the same script.
The worker that runs the Python script is known as the *driver* of the job.

There are three ways to run a Ray job on a Ray cluster:

1. (Recommended) Submit the job using the :ref:`Ray Jobs API <jobs-overview>`.
2. Run the driver script directly on any node of the Ray cluster.

> 3. Use :ref:`Ray Client <ray-client-ref>` to connect remotely to the cluster within a driver script, for interactive development.

Reviewer: I think we should move `, for interactive development.` to bullet point 2 since we want to de-emphasize ray client.

It is important to understand that the autoscaler only reacts to task and actor resource requests, and not application metrics or physical resource utilization.
To learn more about autoscaling, refer to the user guides for Ray clusters on :ref:`VMs <cloud-vm-index>` and :ref:`Kubernetes <kuberay-index>`.

.. _ray_driver:

> Ray Driver

Reviewer 2: Why is this not a part of "running_applications" section? I feel like the right structure is

1. Key concept explains this
2. Under running applications, we should demonstrate (with example) 3 ways to run ray workloads to the ray cluster. 1. driver. 2. job submission. 3. ray client?

Author: Addressed, defined driver inline

----------

A Ray driver is the process that calls the Ray tasks and actors in the user application.

> A driver is started by running `ray.init()`. The script that calls `ray.init()` may also be referred to as the "driver script".

Reviewer 3: I think it would be better to define Ray driver in the same subsection as "Ray job", since we are trying to elevate jobs not drivers.

Also, it is not exactly accurate that the driver needs to call `ray.init()`, since this is now called automatically for certain Ray APIs, like f.remote().

Reviewer: The glossary draft definitions is not correct. Will fix it.

.. _cluster-clients-and-jobs:

> Ray Job

Reviewer: What are the names of the two types of jobs. One is `Ray Job Submission` and what's the other one?

Reviewer: Yea, then do we need a noun called `Ray Job Submission` if it's just a way to run a job.

Or should it be `Submission Job` instead of `Job Submission` to make it clear that it's still a job.

Reviewer 4: Ray job submission -> the submission of the job. It depicts this behavior instead of the job itself.
To sum up,
- There is only 1 job concept: Ray job.
- There are 3 ways (run driver on any node, ray client, submit jobs via job api) to run a Ray job
- There is one api called Job API for submitting jobs to remote clusters, listing jobs, retrieving logs, etc.

Is my understanding aligned with yours? @rkooo567 @architkulkarni @stephanie-wang @jjyao

Reviewer 2: yeah this sounds good.

Author: This sounds good to me.  I'll update this PR accordingly

Reviewer: Yea, that aligns with my understanding as well.

-------

> A Ray Job is a user application that includes a Ray driver. There are two ways of running Ray Jobs on Ray clusters:

Reviewer 2: I am a little confused about this part. I thought we decided to call both driver & submission job as "jobs"? cc @scottsun94

Author: That's what was intended here.  Or are you talking about the fact that it says "Ray Job" instead of "job"?

Author: Submission jobs also contain a ray driver.  Technically they don't have to (the entrypoint could just be `echo hello` instead of `python my_ray_script.py`), but it seems like an edge case and I'd rather not confuse the reader here

Reviewer: I feel @scottsun94's understanding is more clear:

Submission is not a type of job, it's just one way to run the job. so we only have one Job concept but multiple ways to run it.

Reviewer 2: I think I am happy with Stephanie's definition. Btw isn't this 3 ways, not 2 ways? (job submission, driver, and client )

Reviewer: yea, can be 3 ways including ray client.

Reviewer 4: 3 ways are fine too. I counted Driver and Client as the interactive way of running ray jobs above.

> A Ray Job is a user application that includes a Ray driver. There are two ways of running Ray Jobs on Ray clusters:

Reviewer 3: This definition is not very precise? It makes it sound like there are user applications that do not include Ray drivers.

Suggested (adapted from whitepaper):
A Ray job is a single application: it is the collection of Ray tasks, objects, and actors that originate from the same entrypoint. The worker that runs the Python entrypoint is known as the *driver* of the job.

Reviewer 2: Maybe we can call it script?

Reviewer 4: script is clearer to me.

Author: Thanks, I've added the new definition

> The recommended method for running a workload on a Ray cluster is to use Ray Job Submission.

Reviewer 2: Can you add a link to the Job Submission page?

> The recommended method for running a workload on a Ray cluster is to use Ray Job Submission.

Reviewer 2: Actually why don't we structure this way?

```
- [Recommended] Job submission <link>: Blah blah
- Driver: Blah blah
- Client: Blah blah
```

> Ray Job Submission enable users to submit locally developed-and-tested applications to a

Reviewer 3: I probably wouldn't capitalize Submission (lowercasing "jobs" also seems better but we can leave it for later).

```suggestion
Ray Job submission enables users to submit locally developed and tested applications to a
```

remote Ray cluster. Ray Job Submission simplifies the experience of packaging,
deploying, and managing a Ray application.

An alternative method is to run Ray Jobs interactively.  For interactive development, the following methods are available:

* Directly running a :ref:`Ray driver <ray-driver>` Ray driver on any head or worker node.
* Using the Ray Client to connect remotely to the cluster within a :ref:`Ray driver <ray-driver>` script.

> For details about all of these workflows, refer to the :ref:`Ray Jobs guide <jobs-overview>`.

Reviewer 3: ```suggestion
For details on these workflows, refer to the :ref:`Ray Jobs guide <jobs-overview>`.
```
